doc/developer/chips/18.rst

```diff
@@ -0,0 +1,825 @@
+  do not localize execution and may not do so for some time if ever,
+  given the existing limitations on iterator inlining in the compiler
+  plus the fact that changing execution placement is expensive.  So for
+  NUMA performance, parallel iteration is key.
+
+* array element access speed
+
+  The ideal emitted code for an array reference is a memory reference
+  through a (possibly scaled) offset from a base pointer.  Anything more
+  complex than this will perform worse than C or single-ddata Chapel
+  unless it can somehow achieve better affinity to offset that.
+
+* balanced parallelism
+
+  Good load balancing maximizes overall performance in any parallel
+  language construct.  If the parallel construct cannot finish until the
+  task with the most work completes and the work is unbalanced, overall
+  progress is slowed.  The DSIUtil-based domain chunking currently in
+  the ``DefaultRectangular`` domain parallel iterators balances work
+  ideally: the largest and smallest chunks differ by no more than 1 in
+  size.  It seems unlikely that techniques which create significantly
+  unbalanced chunks can achieve NUMA-oriented speedups elsewhere that
+  would offset that imbalance and produce an overall win.
+
+* leveraging existing locality
+
+  Having execution and/or storage locality right in the first place is
+  valuable.  Starting a task takes about the same amount of time no
+  matter where it is placed, but it may take just as long to change
+  where an already existing one is running.  Touching a memory page into
+  existence, which requires storing zeroes into it, costs the same on
+  any NUMA domain but moving a page from one NUMA domain to another
+  takes even longer because the contents are copied even if they are
+  dead (the kernel cannot know).
+
+  As a corollary to the above, where tasks or memory already have NUMA
+  locality and that can be taken advantage of without additional effort
+  it is profitable to do so.  When there is a predefined heap for a
+  Chapel comm layer, for example, if we can arrange NUMA localization in
+  that heap and then use NUMA-aware allocation on it we need not do any
+  additional localization at all.
+
+* network cooperation
+
+  Some network interface chips (NICs), including Cray's Aries, require
+  that memory be "registered" with them in order to be referenced in
+  network operations.  At least for Aries and probably for some others,
+  this involves pinning virtual addresses to physical addresses to fix
+  their relationship.  Doing so necessarily also sets the pages' NUMA
+  locality, since NUMA locality is a physical characteristic.  If the
+  localization of a registered page is to be changed, then the page must
+  be un-registered first and re-registered afterward.  By itself this
+  may not be very costly, but broadcasting the new registration to the
+  other nodes which need it, and synchronizing that broadcast, may be.
```

Comment:
I think it's worth mentioning libfabric here, because it requires all memory involved in data transfers to be registered.